### DISCUSSION TITLE
Does writeFragment require the entity to already be in the cache?

### DISCUSSION THREAD

**Discussion by Axedyson** [urql-graphql/urql#2392] [Q&A] [answered]
This question is related to cache updates using `"@urql/exchange-graphcache"`:

I've tried to use `writeFragment` instead of `updateQuery`, but it seems that `writeFragment` doesn't rerender my react components, but `updateQuery` does. Is the reason for this because `writeFragment` requires entity/data to already be stored in the cache, because if that's the case then fine, I just want to clarify this, thanks!

**Comment by JoviDeCroock** [comment, marked as answer]
What is most likely happening when you create an entity with writeFragment is that it's not linked up to a root-field.

when you update a query like { todo(id: 4) } you automatically write the link of Query.todo to Todo:4 however with writeFragment you still need to use cache.link to tell where this data is accessible. This is the case for creation, for updates this link is already known

**Reply by Axedyson** [reply]
Thanks for the quick response! It works using `cache.link` as you suggested!

**Reply by jdinartejesus** [reply]
@AndysonDK I'm having the same issue, can you share here how you solve it?

**Reply by jdinartejesus** [reply]
@JoviDeCroock it's possible to add connection to other entity from writeFragment?
Eg:

```
    insert_match_schedule_one: (result, { object }, cache) => {
          if (!result?.insert_match_schedule_one) return;

          const match = cache.readFragment(FixtureBaseFragmentDoc, { id: object.match_id });
          const schedules = [...(match?.schedules ?? []), result.insert_match_schedule_one];

          cache.writeFragment(FixtureBaseFragmentDoc, { ...match, schedules, __typename: "match" });
    },
```

**Reply by JoviDeCroock** [reply]
Yes, that should be possible as long as the selection_set and __typename requirements are met

**Reply by jdinartejesus** [reply]
But isn't being added to the cache this schedules, when I try to readFragment after the writeFragment doesn't display the new entry "schedules".

**Reply by JoviDeCroock** [reply]
it's hard for me to tell what's going on from that little code-snippet to be fair 😅 hence I added the "if requirements are met"

**Reply by jdinartejesus** [reply]
What would be the requirements? does the `schedules` needs to be on the FixtureBaseFragmentDoc in order to be allowed to add using  writeFragment?

**Reply by JoviDeCroock** [reply]
yes so for instance

```
... on Whatever {
  id
  __typename
  schedules {
    id
    __typename
  }
}
```

Would be a minimal requirement to write the type and identifier. Do double check whether `result?.insert_match_schedule_one` carries the typename ;)

**Reply by jdinartejesus** [reply]
I see, but in this case it does have the id and a __typename:

<img width="1440" alt="image" src="https://user-images.githubusercontent.com/6892132/170072612-da1f6f09-34f6-417a-89dc-968ea938c3f4.png">

**Reply by JoviDeCroock** [reply]
And does it match all the fields that are in the cache and queried of the schedule or is this resulting in partial data? I don't know whether this is the ideal location to discuss this without a reproduction. There's only so much I can give feedback for without having a common ground of understanding of what you are trying to achieve etc

**Reply by jdinartejesus** [reply]
it's hard to create a codepen to reproduce, there's other place to reach you, so I could have your input?  I'm pretty sure it's a 10 min thing with someone with the experience.

**Reply by JoviDeCroock** [reply]
If you ping me on Twitter or mail me, I would be happy to  setup a call 😄

**Reply by jdinartejesus** [reply]
Thanks! I pinged you on Twitter already 😄
